controllers/tc000243_remediation_retry_test.go

```diff
@@ -0,0 +1,225 @@
+					Reason:  c.Reason,
+					Message: c.Message,
+					Status:  c.Status,
+				}
+			}
+		}
+		return createdHelloWorldTF.Status
+	}, timeout, interval).Should(Equal(expected))
+	g.Expect(createdHelloWorldTF.Status.ReconciliationFailures).To(Equal(int64(3)))
+
+	reconcileResult, err := reconciler.Reconcile(ctx, controllerruntime.Request{
+		NamespacedName: types.NamespacedName{
+			Namespace: createdHelloWorldTF.GetNamespace(),
+			Name:      createdHelloWorldTF.GetName(),
+		},
+	})
+	g.Expect(err).To(Succeed())
+	g.Expect(reconcileResult.Requeue).To(BeFalse())
+
+	recheckHelloWorldTF := infrav1.Terraform{}
+	g.Eventually(func() interface{} {
+		return k8sClient.Get(ctx, helloWorldTFKey, &recheckHelloWorldTF)
+	}, timeout, interval).Should(Succeed())
+
+	var retryCondition *metav1.Condition
+	for _, cond := range recheckHelloWorldTF.Status.Conditions {
+		if cond.Type == meta.StalledCondition && cond.Reason == infrav1.RetryLimitReachedReason {
+			retryCondition = &cond
+			break
+		}
+	}
+	var originalRetryCondition *metav1.Condition
+	for _, cond := range createdHelloWorldTF.Status.Conditions {
+		if cond.Type == meta.StalledCondition && cond.Reason == infrav1.RetryLimitReachedReason {
+			originalRetryCondition = &cond
+			break
+		}
+	}
+
+	g.Expect(retryCondition).ToNot(BeNil(), "Terraform resource should have retry limit reached status condition")
+	g.Expect(originalRetryCondition).ToNot(BeNil(), "Terraform resource should have retry limit reached status condition")
+	g.Expect(recheckHelloWorldTF.Status.ReconciliationFailures).To(Equal(createdHelloWorldTF.Status.ReconciliationFailures))
+	g.Expect(retryCondition.ObservedGeneration).To(Equal(originalRetryCondition.ObservedGeneration))
+	g.Expect(recheckHelloWorldTF.Status.LastAttemptedRevision).To(Equal(testRepo.Status.Artifact.Revision))
+
+	It("should restart retry count")
+	By("when resource was updated")
+	// After changing the resource, retry count should be set back to 0.
+	// With setting Retries lower than the previous one, we can check if it was
+	// reset to 0 as it would never reach 2 from 3.
+	createdHelloWorldTF.Spec.Remediation.Retries = 2
+	g.Expect(k8sClient.Update(ctx, &createdHelloWorldTF)).Should(Succeed())
+
+	expected = failedStatusCheckResult{
+		Type:    meta.StalledCondition,
+		Reason:  infrav1.RetryLimitReachedReason,
+		Message: "Resource reached maximum number of retries.",
+		Status:  metav1.ConditionTrue,
+		Retries: 2,
+	}
+	g.Eventually(func() interface{} {
+		err := k8sClient.Get(ctx, helloWorldTFKey, &createdHelloWorldTF)
+		if err != nil {
+			return nil
+		}
+		for _, c := range createdHelloWorldTF.Status.Conditions {
+			if c.Type == expected.Type && c.Reason == expected.Reason {
+				return failedStatusCheckResult{
+					Type:    c.Type,
+					Reason:  c.Reason,
+					Message: c.Message,
+					Status:  c.Status,
+					Retries: createdHelloWorldTF.Status.ReconciliationFailures,
+				}
+			}
+		}
+		return createdHelloWorldTF.Status
```

Comment:
Returning the `.Status` muddies the water a bit -- it's a value that fails the loop, but it's not `nil`, so ... why return it? (I'll presume copypasta)

Reply:
Yes, it should be `nil` here as well, no reason to return with any values.